core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregationBuilder.java

```diff
@@ -21,38 +21,140 @@
 
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.search.aggregations.AggregatorFactories.Builder;
 import org.elasticsearch.search.aggregations.AggregatorFactory;
 import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValueType;
 import org.elasticsearch.search.aggregations.support.ValuesSource.Numeric;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory;
 import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class HistogramAggregationBuilder extends AbstractHistogramBuilder<HistogramAggregationBuilder> {
     public static final String NAME = InternalHistogram.TYPE.name();
     public static final ParseField AGGREGATION_NAME_FIELD = new ParseField(NAME);
 
+    private double interval;
+    private double offset = 0;
+    private double minBound = Double.MAX_VALUE;
+    private double maxBound = Double.MIN_VALUE;
+
     public HistogramAggregationBuilder(String name) {
-        super(name, InternalHistogram.HISTOGRAM_FACTORY);
+        super(name, InternalHistogram.TYPE, ValuesSourceType.NUMERIC, ValueType.DOUBLE);
     }
 
     /**
      * Read from a stream.
      */
     public HistogramAggregationBuilder(StreamInput in) throws IOException {
-        super(in, InternalHistogram.HISTOGRAM_FACTORY);
+        super(in, InternalHistogram.TYPE, ValuesSourceType.NUMERIC, ValueType.DOUBLE);
+        interval = in.readDouble();
+        offset = in.readDouble();
+        minBound = in.readDouble();
+        maxBound = in.readDouble();
     }
 
     @Override
-    protected HistogramAggregatorFactory innerBuild(AggregationContext context, ValuesSourceConfig<Numeric> config,
-            AggregatorFactory<?> parent, Builder subFactoriesBuilder) throws IOException {
-        return new HistogramAggregatorFactory(name, type, config, interval, offset, order, keyed, minDocCount, extendedBounds, context,
-                parent, subFactoriesBuilder, metaData);
+    protected void innerWriteTo(StreamOutput out) throws IOException {
+        super.innerWriteTo(out);
+        out.writeDouble(interval);
+        out.writeDouble(offset);
+        out.writeDouble(minBound);
+        out.writeDouble(maxBound);
+    }
+
+    public double interval() {
+        return interval;
+    }
+
+    public HistogramAggregationBuilder interval(double interval) {
+        if (interval <= 0) {
+            throw new IllegalArgumentException("[interval] must be >0 for histogram aggregation [" + name + "]");
+        }
+        this.interval = interval;
+        return this;
+    }
+
```

Comment:
This is a big antipattern: here we have two methods, both named interval(), just with different signatures. One is actually a getter, whereas the other changes state!

But you have no idea reading source code, what is going on, e.g. its just a method and you have to follow the source code to figure out what is happening. 

 Please change to getInterval() and setInterval() [there are other instances of these in these files, obviously those should be fixed, too]